server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,745 @@
+
+        NamedAnalyzer normalizer = fieldType().normalizer();
+        if (normalizer != null) {
+            value = normalizeValue(normalizer, name(), value);
+        }
+
+        String[] valueTypeList = fieldName.split("\\._");
+        String valueType = "._" + valueTypeList[valueTypeList.length - 1];
+
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
+            // convert to utf8 only once before feeding postings/dv/stored fields
+
+            final BytesRef binaryValue = new BytesRef(fieldType().name() + DOT_SYMBOL + value);
+            Field field = new FlatObjectField(fieldType().name(), binaryValue, fieldType);
+
+            if (fieldType().hasDocValues() == false && fieldType.omitNorms()) {
+                createFieldNamesField(context);
+            }
+            if (fieldName.equals(fieldType().name())) {
+                context.doc().add(field);
+            }
+            if (valueType.equals(VALUE_SUFFIX)) {
+                if (valueFieldMapper != null) {
+                    valueFieldMapper.addField(context, value);
+                }
+            }
+            if (valueType.equals(VALUE_AND_PATH_SUFFIX)) {
+                if (valueAndPathFieldMapper != null) {
+                    valueAndPathFieldMapper.addField(context, value);
+                }
+            }
+
+            if (fieldType().hasDocValues()) {
+                if (context.doc().getField(fieldType().name()) == null || !context.doc().getFields(fieldType().name()).equals(field)) {
+                    context.doc().add(new SortedSetDocValuesField(fieldType().name(), binaryValue));
+                }
+            }
+
+        }
+
+    }
+
+    private static String normalizeValue(NamedAnalyzer normalizer, String field, String value) throws IOException {
+
+        try (TokenStream ts = normalizer.tokenStream(field, value)) {
+            final CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+            ts.reset();
+            if (ts.incrementToken() == false) {
+                throw new IllegalStateException(
+                    "The normalization token stream is "
+                        + "expected to produce exactly 1 token, but got 0 for analyzer "
+                        + normalizer
+                        + " and input \""
+                        + value
+                        + "\""
+                );
+            }
+            final String newValue = termAtt.toString();
+            if (ts.incrementToken()) {
+                throw new IllegalStateException(
+                    "The normalization token stream is "
+                        + "expected to produce exactly 1 token, but got 2+ for analyzer "
+                        + normalizer
+                        + " and input \""
+                        + value
+                        + "\""
+                );
```

Comment:
can we reuse the error message string?

Reply:
resolved in new commit [Add more IT tests for supported queries](https://github.com/opensearch-project/OpenSearch/pull/6507/commits/cdf6797de8930613e187101e3032285dee76df19)